ts/sagas/startup.ts

```diff
@@ -316,69 +316,57 @@ export function* initializeApplicationSaga(): Generator<
   // Start watching for requests of abort the onboarding
   const watchAbortOnboardingSagaTask = yield* fork(watchAbortOnboardingSaga);
 
-  if (!previousSessionToken || O.isNone(maybeStoredPin)) {
-    // The user wasn't logged in when the application started or, for some
-    // reason, he was logged in but there is no unlock code set, thus we need
-    // to pass through the onboarding process.
+  const hasPreviousSessionAndPin =
+    previousSessionToken && O.isSome(maybeStoredPin);
+  if (hasPreviousSessionAndPin) {
+    // we ask the user to identify using the unlock code.
+    // FIXME: This is an unsafe cast caused by a wrongly described type.
+    const identificationResult: SagaCallReturnType<
+      typeof startAndReturnIdentificationResult
+    > = yield* call(startAndReturnIdentificationResult, maybeStoredPin.value);
+
+    if (identificationResult === IdentificationResult.pinreset) {
+      // If we are here the user had chosen to reset the unlock code
+      yield* put(startApplicationInitialization());
+      return;
+    }
+  }
 
-    // Ask to accept ToS if it is the first access on IO or if there is a new available version of ToS
-    yield* call(checkAcceptedTosSaga, userProfile);
+  // Ask to accept ToS if there is a new available version
+  yield* call(checkAcceptedTosSaga, userProfile);
 
-    // check if the user expressed preference about mixpanel, if not ask for it
-    yield* call(askMixpanelOptIn);
+  // check if the user expressed preference about mixpanel, if not ask for it
+  yield* call(askMixpanelOptIn);
 
+  if (hasPreviousSessionAndPin) {
+    // We have to retrieve the pin here and not on the previous if-condition (same guard)
+    // otherwise the typescript compiler will complain of an unassigned variable later on
+    storedPin = maybeStoredPin.value;
+  } else {
+    // TODO If the session was not valid, the code would have stopped before
+    // reaching this point. Consider refactoring even more by removing the check
+    // on the session (IOAPPCIT-10)
     storedPin = yield* call(checkConfiguredPinSaga);
 
     yield* call(checkAcknowledgedFingerprintSaga);
 
     yield* call(checkAcknowledgedEmailSaga, userProfile);
+  }
```

Comment:
Could this block of code be moved inside the `if` clause at line [321](https://github.com/pagopa/io-app/pull/4168/files#diff-8a5b2f3967d681b976fe673762bd1061f5b430130c880c1195b76af06362cf31R321)?